clippy_lints/src/needless_pass_by_ref_mut.rs

```diff
@@ -289,14 +323,38 @@ impl<'tcx> euv::Delegate<'tcx> for MutablyUsedVariablesCtxt {
             {
                 self.add_mutably_used_var(*vid);
             }
+        } else if borrow == ty::ImmBorrow {
+            // If there is an `async block`, it'll contain a call to a closure which we need to
+            // go into to ensure all "mutate" checks are found.
+            if let Node::Expr(Expr {
+                kind:
+                    ExprKind::Call(
+                        _,
+                        [
+                            Expr {
+                                kind: ExprKind::Closure(Closure { def_id, .. }),
+                                ..
+                            },
+                        ],
+                    ),
+                ..
+            }) = self.tcx.hir().get(id)
```

Comment:
Sorry I'm a bit lost: what is `diag_expr_id` here? What's the connection with `consume` too?

In this case, `async` blocks generate a closure and an `Expr` containing a call to a `closure` which has a `DefId` and that's what I retrieve with this code. If this is changed in the compiler, the UI test will catch it.

Reply:
`diag_expr_id` i.e., `id`. That's the default name for it.

Its intended use is clarified in `Delegate::consume`'s docs, from what I can tell it's not supposed to be used for analysis.

Reply:
Oh ok I see. In this case, it's all generated by the compiler (it doesn't exist in user code). The case of `async block` seems to be special if you look at [ExprKind::Closure](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/enum.ExprKind.html#variant.Closure). So what do you want to do in this case?

Reply:
It's probably fine as is, but once we can we should change it. For now however, we could change this to `cmt.hir_id` which would convey this better (`diag_expr_id` is currently always `cmt.hir_id` for `borrow`) as `cmt` (`place_with_id`) is usually where the binding was introduced I believe (in this case, the closure, though it would realistically be the parameters' `HirId`)

Reply:
That makes perfect sense, thanks for the detailed explanation!